src/libraries/System.Private.CoreLib/src/System/DateTime.Windows.cs

```diff
@@ -1,30 +1,164 @@
+                // by +1 or -1 second, and only at the very end of the day. That is, time rolls
+                // 23:59:58 -> 00:00:00 (negative leap second) or 23:59:59 -> 23:59:60 (positive leap
+                // second). Thus we assume that each day has at least 23 hr 59 min 59 sec, or
+                // 863,990,000,000 ticks.
+                //
+                // We take advantage of this by caching what the OS believes the tick count is at
+                // the beginning of the month vs. what .NET believes the tick count is at the beginning
+                // of the month. When the OS returns a tick count to us, if it's before 23:59:59 on the
+                // last day of the month, then we know that there's no way for a leap second to have been
+                // inserted or removed, and we can short-circuit the leap second handling logic by
+                // performing a quick addition and returning immediately.
+                //
+                // This relies on:
+                // a) Leap seconds only ever taking place on the last day of the month
+                //    (historically these are Jun. & Dec., but we support any month);
+                // b) Leap seconds only ever occurring at 23:59:59;
+                // c) At most a single leap second being added or removed per month; and
+                // d) Windows never inserting a historical leap second (into the past) once the
+                //    process is up and running (future insertions are ok).
+
+                // TODO: Once https://github.com/dotnet/runtime/issues/38134 is resolved,
+                // the call below should suppress the GC transition.
+
+                ulong osTicks;
+                s_pfnGetSystemTimeAsFileTime(&osTicks);
+
+                // If the OS doesn't support leap second handling, short-circuit everything.
+
+                if (!s_systemSupportsLeapSeconds)
                 {
-                    FullSystemTime time;
-                    GetSystemTimeWithLeapSecondsHandling(&time);
-                    return CreateDateTimeFromSystemTime(in time);
+                    return new DateTime((osTicks + FileTimeOffset) | KindUtc);
                 }
 
-                return new DateTime(((ulong)(GetSystemTimeAsFileTime() + FileTimeOffset)) | KindUtc);
+                // If our cache has been populated and we're within its validity period, rely
+                // solely on the cache. We use unsigned arithmetic below so that if the system
+                // clock is set backward, the resulting value will integer underflow and cause
+                // a cache miss, forcing us down the slow path.
+
+                if (s_leapSecondCache is LeapSecondCache cache)
+                {
+                    ulong deltaTicksFromStartOfMonth = osTicks - cache.OSFileTimeTicksAtStartOfMonth;
+                    if (deltaTicksFromStartOfMonth < cache.CacheValidityPeriodInTicks)
+                    {
```

Comment:
I know we talked offline about that and you mentioned the line 
```C#
ulong deltaTicksFromStartOfMonth = osTicks - cache.OSFileTimeTicksAtStartOfMonth;
```

will get optimized. but I am not sure why we need to calculate the delta at all every time. we can just get rid of this delta calculation and calculate it once when we are caching it. 

When caching, we calculate the delta of ticks = OSTicks - (dotnet calculated ticks from osticks)  // let's call it ticksDelta
Also, when caching we'll calculate the OS ticks at the end of the current month at 23:59:59       // let's call it endOfMonthOSTicks

the code will be simplified as follow:
```C#
        if (s_leapSecondCache is LeapSecondCache cache && cache.endOfMonthOSTicks > osticks)
        {
             return new DateTime(osticks - ticksDelta + FileTimeOffset); // includes UTC marker
        }
```

Reply:
I don't think this optimization would be valid. Consider the case where my system clock is Jan 1, 2017 00h00:01 UTC. I call `DateTime.UtcNow` and populate the cache. Then Windows syncs the system clock against the network time server, and it turns out that my machine was oh-so-slightly fast. So I go back 3 seconds, and now my system clock is Dec 31, 2016 23h59:58 UTC per .NET's reckoning. But Dec 31, 2016 had a positive leap second! So 3 seconds behind Jan 1, 2017 00h00:01 UTC should actually be reported as Dec 31, 2016 23h59:59 UTC.

Is there a good way to account for this condition other than including both the start of month and the end of month times in the cache?

Reply:
looks this scenario will be broken with your code too. no? you can run into the same issue when calculating OSFileTimeTicksAtStartOfMonth too. right? I think if the time on the system can be adjusted at any time we'll break as we'll not feel this adjustment.

Reply:
I don't think this scenario is broken with the proposed implementation. If the clock gets set backward to the previous month, then `GetSystemTimeAsFileTime` will return a tick count _less than_ our cached value `cache.OSFileTimeTicksAtStartOfMonth`, and _deltaTicksFromStartOfMonth_ will end up being a very large positive number (due to integer overflow). This causes the `deltaTicksFromStartOfMonth < cache.CacheValidityPeriodInTicks` check to fail, sending us down the fallback code path.

Reply:
I understand cannot insert historical leap seconds but you can still insert leap seconds at the end of the list (I guess). meaning you can start caching, then leap second get insert at the end of previous month (and would be the last leap second in the list for Windows). I think this will break this caching logic. you can use the tool I pointed you at for inserting leap second and then you can try this scenario and look what you'll get.